src/main/java/io/aiven/kafka/connect/http/converter/RecordValueConverter.java

```diff
@@ -16,36 +16,99 @@
-            HashMap.class, jsonRecordValueConverter,
-            LinkedHashMap.class, jsonRecordValueConverter,
-            Struct.class, jsonRecordValueConverter
+        String.class, record -> (String) record.value(),
+        Map.class, jsonRecordValueConverter,
+        Struct.class, jsonRecordValueConverter
     );
 
     interface Converter {
         String convert(final SinkRecord record);
     }
 
     public String convert(final SinkRecord record) {
-        if (!converters.containsKey(record.value().getClass())) {
+        final Converter converter = getConverter(record);
+        return converter.convert(record);
+    }
+
+    private Converter getConverter(final SinkRecord record) {
+        return RUNTIME_CLASS_TO_CONVERTER_CACHE.computeIfAbsent(record.value().getClass(), clazz -> {
+            final boolean directlyConvertible = converters.containsKey(clazz);
+            final List<Class<?>> convertibleByImplementedTypes = getAllSerializableImplementedInterfaces(clazz);
+            validateConvertibility(clazz, directlyConvertible, convertibleByImplementedTypes);
+
+            Class<?> implementedClazz = clazz;
+            if (!directlyConvertible) {
+                implementedClazz = convertibleByImplementedTypes.get(0);
+            }
+            return converters.get(implementedClazz);
+        });
+    }
+
+    private List<Class<?>> getAllSerializableImplementedInterfaces(final Class<?> recordClazz) {
+        // caching the computation since querying implemented interfaces is expensive.
+        // The size of the cache is unlimited, but I don't think it's a problem
+        // since the number of different record classes is limited.
+        return getAllInterfaces(recordClazz).stream()
+            .filter(converters::containsKey)
+            .collect(Collectors.toList());
+    }
+
+    public static Set<Class<?>> getAllInterfaces(final Class<?> clazz) {
+        final Set<Class<?>> interfaces = new HashSet<>();
+
+        for (final Class<?> implementation : clazz.getInterfaces()) {
+            interfaces.add(implementation);
+            interfaces.addAll(getAllInterfaces(implementation));
+        }
+
+        if (clazz.getSuperclass() != null) {
+            interfaces.addAll(getAllInterfaces(clazz.getSuperclass()));
+        }
+
+        return interfaces;
+    }
+
+    private static void validateConvertibility(
+        final Class<?> recordClazz,
+        final boolean directlyConvertible,
+        final List<Class<?>> convertibleByImplementedTypes
+    ) {
+        final boolean isConvertibleType = directlyConvertible || !convertibleByImplementedTypes.isEmpty();
+
+        if (!isConvertibleType) {
+            throw new DataException(
+                String.format(
+                    "Record value must be a String, a Schema Struct or implement one of "
+                        + "[`java.util.Map.Entry`, `java.util.Map`] interfaces, but %s is given",
```

Comment:
Can it really be a Map.Entry? Seems it only supports String, Map and Struct no?

Reply:
oh right, forgot from previous implementation! Thanks